container: add support for libkrun

giuseppe opened the pull request:
add a custom annotation to specify whether the container must be
executed using libkrun.

crun must be configured and compiled with `--with-libkrun`.
    
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

example usage:
```
$ podman run -v /dev/kvm:/dev/kvm  --annotation run.oci.handler=krun --rm fedora echo hello
hello
```

giuseppe commented:
rebased on top of https://github.com/containers/crun/pull/438 as some changes introduced by https://github.com/containers/crun/pull/438 are required here

giuseppe commented:
@slp @rhatdan FYI

rhatdan commented:
Very cool
LGTM

giuseppe commented:
@slp, do you foresee any breaking change in the libkrun APIs so that it is worth to wait before merging this PR?

slp commented:
> @slp, do you foresee any breaking change in the libkrun APIs so that it is worth to wait before merging this PR?

@giuseppe Actually, I'm not happy with the current API, as it's going to make very hard to dealing with future updates. I'm thinking about replacing it with something like this:

```
// Returns library version on success, or -1 on error
int krun_init(void);
// Sets log level
int krun_set_log_level(uint8_t level);
// Sets the number of vCPUs
int krun_set_num_vcpus(uint8_t num_vcpus);
(...)
// Initialize AMD SEV environment and load the initial payload
int krun_sev_init(struct sev_policy, const char *cert_chain, size_t cert_chain_size);
// Retrieve SEV launch measurement
int krun_sev_get_measurement(char **measurement, size_t *measurement_size);
(...)
// Enter the VM and don't return until it has exited.
int krun_enter();
```

Newer versions of the library would add new symbols, but never remove nor modify the existing ones. Library users can check the version returned by `krun_init` to verify the set of supported features. How does this look from your perspective?

BTW, I'm working on adding SEV support so users can run encrypted workloads and, ideally, there should be a protocol that would allow podman/CRI-O (or some entity controlling them) to verify the measurement of the payload for attestation. What's the proper place for this kind of discussions?

rhatdan commented:
@giuseppe Needs a rebase.
I think you should cut a release before we merge this.

rhatdan commented:
@slp @giuseppe I would love to get this into Fedora so that people could start playing with it.  We would need some video blogs on how to do this with Podman.  What packages need to be installed, how to setup a kernel, although in a perfect world this could happen automatically or be packaged up with the host kernel (Not likely)

giuseppe commented:
> How does this look from your perspective?

it works fine for crun

> BTW, I'm working on adding SEV support so users can run encrypted workloads and, ideally, there should be a protocol that would allow podman/CRI-O (or some entity controlling them) to verify the measurement of the payload for attestation. What's the proper place for this kind of discussions?

how much of the logic must end up in podman/CRI-O and how much in the OCI runtime?  If most of the implementation must be done before crun runs, we could have a discussion on the podman mailing list.  Otherwise if the implementation is mostly in the OCI runtime/libkrun layer, I guess we could even have it here?

rhatdan commented:
@giuseppe @slp Needs a rebase.

slp commented:
> @slp @giuseppe I would love to get this into Fedora so that people could start playing with it. We would need some video blogs on how to do this with Podman. What packages need to be installed, how to setup a kernel, although in a perfect world this could happen automatically or be packaged up with the host kernel (Not likely)

I'd like to mature `libkrun` a bit before attempting to get it into Fedora. But I also had the intention of creating a `COPR` repo as soon this PR gets merged, including `crun` (built with `--with-libkrun` and shipped with a different binary name, so it can be installed alongside the original one), `libkrun` and `libkrunfw`, so users can test it easily.

slp commented:
> > How does this look from your perspective?
> 
> it works fine for crun. So I'll keep this PR as a draft until we settle on the API :-)

This week I'll focus on updating the API.

> > BTW, I'm working on adding SEV support so users can run encrypted workloads and, ideally, there should be a protocol that would allow podman/CRI-O (or some entity controlling them) to verify the measurement of the payload for attestation. What's the proper place for this kind of discussions?
> 
> how much of the logic must end up in podman/CRI-O and how much in the OCI runtime? If most of the implementation must be done before crun runs, we could have a discussion on the podman mailing list. Otherwise if the implementation is mostly in the OCI runtime/libkrun layer, I guess we could even have it here?

Actually, it's quite a wide topic, as the upper layers also need to participate in the exchange of information. I guess I'll initiate the discussion on the podman mailing list, and we'll see where that leads us.

slp commented:
@giuseppe I've created a PR in libkrun with a new API that should allow us to evolve the library while retaining backwards compatibility. If possible, please take a look a it to check that it'll play nice with crun. If you find something you don't like or that you think it could be done in a better way, please don't hesitate to tell me:

- https://github.com/containers/libkrun/blob/2433c0b70de75af85dbb123b23ca164a5e581d1f/include/libkrun.h
- https://github.com/containers/libkrun/pull/5

Thanks!

giuseppe commented:
>  If you find something you don't like or that you think it could be done in a better way,

do you think it would make sense to have a way to configure and use different contexts?  e.g. `int32_t krun_init(krun_context *ctx);` and `int32_t krun_set_root(krun_context *ctx, const char *root_path);`

With the current API, it looks like it is possible to configure just one VM per process.  Not sure if it is technically possible now, but at least we won't block this possibility in the API.

giuseppe commented:
>  If you find something you don't like or that you think it could be done in a better way,

do you think it would make sense to have a way to configure and use different contexts?  e.g. `int32_t krun_init(krun_context *ctx);` and `int32_t krun_set_root(krun_context *ctx, const char *root_path);`

With the current API, it looks like it is possible to configure just one VM per process.  Not sure if it is technically possible now, but at least we won't block this possibility in the API.

slp commented:
> > If you find something you don't like or that you think it could be done in a better way,
> 
> do you think it would make sense to have a way to configure and use different contexts? e.g. `int32_t krun_init(krun_context *ctx);` and `int32_t krun_set_root(krun_context *ctx, const char *root_path);`
> 
> With the current API, it looks like it is possible to configure just one VM per process. Not sure if it is technically possible now, but at least we won't block this possibility in the API.

I was planning to tackle multi-VM support in the future (mainly, because I don't see a clear use case), but let's see if we can at least prepare the API for that, as you suggested. Thanks!

slp commented:
> > > If you find something you don't like or that you think it could be done in a better way,
> > 
> > 
> > do you think it would make sense to have a way to configure and use different contexts? e.g. `int32_t krun_init(krun_context *ctx);` and `int32_t krun_set_root(krun_context *ctx, const char *root_path);`
> > With the current API, it looks like it is possible to configure just one VM per process. Not sure if it is technically possible now, but at least we won't block this possibility in the API.
> 
> I was planning to tackle multi-VM support in the future (mainly, because I don't see a clear use case), but let's see if we can at least prepare the API for that, as you suggested. Thanks!

@giuseppe Done! PTAL, thanks!

https://github.com/containers/libkrun/blob/5ecfa5caf97663350ffea3fae8d1b0adaa6e0a8d/include/libkrun.h

giuseppe commented:
> > 
> 
> @giuseppe Done! PTAL, thanks!
> 
> https://github.com/containers/libkrun/blob/5ecfa5caf97663350ffea3fae8d1b0adaa6e0a8d/include/libkrun.h

thanks!

Just one suggestion: do you think `int32_t krun_create_ctx(...);` should create the context_id itself?  So that the context_id is the return value instead of accepting it as input.

For API completeness, even if a NO-OP for now, could we have a `krun_free_ctx(..)`?

slp commented:
> > > 
> > 
> > 
> > @giuseppe Done! PTAL, thanks!
> > https://github.com/containers/libkrun/blob/5ecfa5caf97663350ffea3fae8d1b0adaa6e0a8d/include/libkrun.h
> 
> thanks!
> 
> Just one suggestion: do you think `int32_t krun_create_ctx(...);` should create the context_id itself? So tht the context_id is the return value instead of accepting it as input.
> 
> For API completeness, even if a NO-OP for now, could we have a `krun_free_ctx(..)`?

Done and done, PTAL, thanks! :-)

rhatdan commented:
@slp Looks like we need a rebase.

giuseppe commented:
> Done and done, PTAL, thanks! :-)

thanks!  The API looks good

giuseppe commented:
rebased on top of the new API.

@slp, if we can consider the API "stable" I am fine with merging this PR, as anyway it is enabled via a configure option (disabled by default)

slp commented:
> rebased on top of the new API.
> 
> @slp, if we can consider the API "stable" I am fine with merging this PR, as anyway it is enabled via a configure option (disabled by default)

@giuseppe Yes, this API will be stable. New functionality will be added as new symbols.

It'd be great to have this merged. As soon it's there, I'll create a COPR repo with crun (installed on a different location) and libkrun so people can easily give it a try.

giuseppe commented:
moved to ready for review.

One issue I am seeing is with the `nginx` image, the port is not accessible when running with krun:

```
podman run --annotation run.oci.libkrun=/usr/local/lib64/libkrun.so -p 8080:80 --rm  -v /dev/kvm:/dev/kvm docker.io/nginx:alpine sh
# /docker-entrypoint.sh nginx
```

If I don't specify a different command `sh`, the container exits immediately

rhatdan commented:
It would be interesting to attempt to run some of the system tests of podman with libkrun, to see how it does.

slp commented:
> moved to ready for review.
> 
> One issue I am seeing is with the `nginx` image, the port is not accessible when running with krun:

Seems like `nginx` doesn't like some aspect of the `s2vi` (socket-2-vsock impersonation) mechanism, which is not surprising. I'll debug it later this week. The simple HTTP server embedded into python (`python3 -m http.server --bind 0.0.0.0`) seems to work fine.
 
> If I don't specify a different command `sh`, the container exits immediately

This happens because `krun_set_exec` expects `args` to be a collapsed string from every argument, instead of a array of string pointers, to avoid doing that in Rust. Now I see this is clearly counterintuitive, so given that we're still in time to fix this in `libkrun`'s public API, I'm going to do it there.

giuseppe commented:
back to Draft.  I'll update again once the API is changed.  Once this is merged I'll cut a new release.

slp commented:
@giuseppe I've updated the API to make `krun_set_exec` receive an array of string pointers for both the arguments and the environment variables lists, in the same fashion as `execve(2)` does.

This implies a couple of changes in this PR, basically updating the signature for `krun_set_exec` and passing `&argv[1]` instead of `argv[1]`, but I think it's worth it.

giuseppe commented:
> @giuseppe I've updated the API to make `krun_set_exec` receive an array of string pointers for both the arguments and the environment variables lists, in the same fashion as `execve(2)` does.
> 
> This implies a couple of changes in this PR, basically updating the signature for `krun_set_exec` and passing `&argv[1]` instead of `argv[1]`, but I think it's worth it.

I've updated the PR to use the new version, thanks!

giuseppe commented:
@slp @rhatdan are you fine with the crun API for using libkrun?

slp commented:
> @slp @rhatdan are you fine with the crun API for using libkrun?

@giuseppe LGTM. 👍

rhatdan commented:
If SLP is fine with this, then I am fine with it.
